### DISCUSSION TITLE
The JIT release

### DISCUSSION THREAD

**Discussion by ben-rogerson** [ben-rogerson/twin.macro#514] [Releases]
This release brings the same special features we all love from the Tailwind JIT release! (v2.2.0-2.2.7) 🎉🌮
As this is the same feature set as Tailwind, you can read some more information about each feature in the [Tailwind release notes](https://github.com/tailwindlabs/tailwindcss/releases/tag/v2.2.0).

### A bunch of new variants

- `peer-*:` variants for styling sibling elements ([PR](https://github.com/ben-rogerson/twin.macro/pull/480))
- `first-letter:` and `first-line:` pseudo-elements ([PR](https://github.com/ben-rogerson/twin.macro/pull/477))
- `selection:` for styling text selections ([PR](https://github.com/ben-rogerson/twin.macro/pull/477))
- `marker:` for styling list markers ([PR](https://github.com/ben-rogerson/twin.macro/pull/477))
- and more: `autofill:`, `in-range:`, `out-of-range:`, `first-letter:`, `first-line:`, `only:`, `empty:`, `selection:`, `marker:`, `rtl:` and `ltr:` ([PR](https://github.com/ben-rogerson/twin.macro/pull/477))

[Check the variant config for a full list →](https://github.com/ben-rogerson/twin.macro/blob/master/src/config/variantConfig.js)

### Shorthand color opacity syntax ([PR](https://github.com/ben-rogerson/twin.macro/pull/485))

Add a slash opacity value to set the opacity on any color classes:

```js
tw`bg-black/50`

// ↓ ↓ ↓ ↓ ↓ ↓

({
  "backgroundColor": "rgba(0, 0, 0, 0.5)"
});
```
This slash value comes from your opacity setting in your tailwind config.
For custom slash opacity values, use square brackets:

```js
tw`bg-black/[.22]`
tw`bg-black/[--my-opacity-variable]`
```

### Extended arbitrary value support ([PR](https://github.com/ben-rogerson/twin.macro/pull/447))

The "dash square bracket" syntax adds custom values on just about every dynamic class (like `bg-red-500`, not `block`):

```js
tw`col-start-[73] placeholder-[#aabbcc] object-[50%]`

// ↓ ↓ ↓ ↓ ↓ ↓

({
  "gridColumnStart": "73",
  "::placeholder": {
    "--tw-placeholder-opacity": "1",
    "color": "rgba(170, 187, 204, var(--tw-placeholder-opacity))"
  },
  "objectPosition": "50%"
});
```

This feature _almost_ replace twin's "[short css](https://github.com/ben-rogerson/twin.macro/releases/tag/2.2.1)" feature but short css is still great for more obscure css as you can specify custom properties too:

```js
tw`--my-var[red]`
tw`-webkit-gradient[gradient here]`
```

### Per-side border color utilities ([PR](https://github.com/ben-rogerson/twin.macro/pull/505))

You can now specify border color and border width on a specific border direction:

```js
// Colors
tw`border-t-gray-50`;
tw`border-b-red-500`;

// ↓ ↓ ↓ ↓ ↓ ↓

({
  "--tw-border-opacity": "1",
  "borderTopColor": "rgba(249, 250, 251, var(--tw-border-opacity))"
});
({
  "--tw-border-opacity": "1",
  "borderBottomColor": "rgba(239, 68, 68, var(--tw-border-opacity))"
});
```
```js
// Widths
tw`border-t-4`;
tw`border-b-4`;

// ↓ ↓ ↓ ↓ ↓ ↓

({
  "borderTopWidth": "4px"
});
({
  "borderBottomWidth": "4px"
}); 
```

## Smaller features

- **`before:` and `after:` variants now automatically add `content: ''`** ([PR](https://github.com/ben-rogerson/twin.macro/pull/479))<br/>No more requirement for the `content` class every time we use pseudo classes
- **Simplified transform and filter composition**<br/>Now there's no need to add `transform` or `filter` to make your _actual_ transform or filter classes work (Possible breaking change - check [Adams PR notes](https://github.com/tailwindlabs/tailwindcss/pull/4604))
- **Add new `transform-cpu` class**<br/>This class uses `translateX` and `translateY` instead of `translate3d` which is added by `transform-gpu`
- **Background origin utilities** (Thanks @theonesean) ([PR](https://github.com/ben-rogerson/twin.macro/pull/502))<br/>Classes for controlling how an element's background is positioned relative to borders, padding, and content - [Docs](https://tailwindcss.com/docs/background-origin)
- **Caret color utilities** ([PR](https://github.com/ben-rogerson/twin.macro/pull/506))<br/>Specify the color of the carat in textareas/inputs/contenteditables (eg: ``tw`caret-red-500` ``)

## Support

If you’re kicking ass with twin - especially if you’re in a team - then please consider [becoming a sponsor](https://github.com/sponsors/ben-rogerson). 
It will really help out with the continued development of twin - cheers! 🍻 

<hr /><em>This discussion was created from the release <a href='https://github.com/ben-rogerson/twin.macro/releases/tag/2.7.0'>The JIT release</a>.</em>

**Comment by u3u** [comment]
Hi, how can I disable `preflight` in version 2.7.0?
In the new version, to make `transform` and `filter` work I had to introduce global styles
Considering exporting all of these for users to freely combine or respect the `corePlugins.preflight` option?

https://github.com/ben-rogerson/twin.macro/blob/ac7b194a9a0b7e1deed1feb5f42c1bff6060020d/src/config/globalStyles.js#L77-L86

**Reply by ben-rogerson** [reply]
I'd have to have a ponder on the best way to allow a "pick-n-choose" of global styles - I'm hesitant to provide more exports as this is a babel macro plugin we're working with and they don't function like normal exports.

Perhaps it could be underneath a custom tailwind.config setting, something like this:

```js
// tailwind.config.js
module.exports = {
  theme: {},
  twin: {
    globals: ({
      modernNormalizeStyles,
      globalPreflightStyles,
      globalKeyframeStyles,
      globalTransformStyles,
      // globalRingStyles,
      // globalBoxShadowStyles,
      // globalFilterStyles,
      // globalBackdropStyles,
    }) => ({
      modernNormalizeStyles,
      globalPreflightStyles,
      globalKeyframeStyles,
      globalTransformStyles,
      // globalRingStyles,
      // globalBoxShadowStyles,
      // globalFilterStyles,
      // globalBackdropStyles,
    }),
  },
}
```

This would also allow custom css to be injected via your tailwind.config.

Right now, you can use [patch-package](https://www.npmjs.com/package/patch-package) to make the necessary adjustments to the twin code _or_ provide a PR with the functionality to help speed this along.
